# || Chapter 09 - 추천시스템 ||

## 01. 추천 시스템의 개요와 배경

1. 추천 시스템의 개요
    - 사용자 자신도 좋아하는지 몰랐던 취향을 시스템이 발견하고 추천해줌
    - 고객을 오래 머무르게 하고 매출 증가


2. 온라인 스토어의 필수 요소, 추천 시스템
    - 좋은 추천 시스템은 사용자가 무엇을 원하는지 빠르게 찾아내 사용자의 온라인 쇼핑 즐거움을 배가함

![image](https://user-images.githubusercontent.com/79856553/151126453-460ca403-f842-4a06-9a30-d50e19977f27.png)

3. 추천 시스템의 유형
    1. **콘텐츠 기반 필터링**
    2. **협업 필터링**
        - **최근접 이웃 협업 필터링**: 아마존 사용
        - **잠재 요인 협업 필터링**: 대부분 온라인 스토어에서 적용중
    - 최근에는 개인화 특성을 좀 더 강화하기 위해 하이브리드 형식으로 콘텐츠 기반과 협업 기반을 적절히 결합해 사용하는 경우도 증가

## 02. 콘텐츠 기반 필터링 추천 시스템

1. **콘텐츠 기반 필터링**: 사용자가 **특정한 아이템을 매우 선호**하는 경우, 그 아이템과 비슷한 콘텐츠를 가진 다른 아이템을 추천하는 방식
    - ex) 사용자가 특정 영화에 높은 평점을 줬다면 그 영화의 장르, 출연 배우, 감독, 영화 키워드 등을 콘텐츠와 유사한 다른 영화 추천

## 03. 최근접 이웃 협업 필터링

1. **협업 필터링**: 친구들에게 물어보는 것과 유사한 방식으로, 사용자가 아이템에 매긴 평점 정보나 상품 구매 이력과 같은 **사용자 행동 양식만을 기반으로** 추천을 수행하는 것


2. 주요 목표: 사용자-아이템 평점 매트릭스와 같은 축적된 **사용자 행동 데이터를 기반으로** 사용자가 아직 평가하지 않은 아이템을 **예측 평가**하는 것


3. **사용자-아이템 평점 행렬**
    - 행(Row): 개별 자용자
    - 열(Column): 개별 아이템
    - 레코드 레벨 형태 -> 평점 행렬 형태 변환(pivot_table())
        - 희소 행렬 특성 가짐

![image](https://user-images.githubusercontent.com/79856553/151132106-53459255-e6f5-4e6e-9913-42974efcc8f2.png)

4. **최근접 이웃 협업 필터링**(=메모리 협업 필터링)
    1. **사용자 기반(User-User)**: 당신과 비슷한 고객들이 다음 상품도 구매
        - 특정 사용자와 타 사용자 간의 유사도 측정: 유사도 높은 TOP-N사용자
        - TOP-N 사용자가 선호하는 아이템 추천
    2. **아이템 기반(Item-Item)**: 이 상품 선택한 다른 고객들은 다음 상품도 구매
        - 사용자들이 그 아이템을 좋아하는지/싫어하는지 평가 척도가 유사한 아이템 추천
        - **사용자 기반보다 정확도 높음**(유명한 영화는 취향과 관계X 모두 관람)
    - 유사도 측정에는 **코사인 유사도** 이용
    $$ similarity = \cos\theta = \frac{\vec{A}\cdot\vec{B}}{||A||||B||} = \frac{\sum_{i=1}^n A_iB_i}{\sqrt{\sum_{i=1}^n A_i^2}\sqrt{\sum_{i=1}^n B_i^2}} $$

![image](https://user-images.githubusercontent.com/79856553/151133176-4dbd39a4-c30d-4a1f-8a0b-aa35d22a07b7.png)

![image](https://user-images.githubusercontent.com/79856553/151133234-e6501f50-0765-4cd7-8de0-843224f6da70.png)

## 04. 잠재 요인 협업 필터링

1. **잠재 요인 협업 필터링**: 사용자-아이템 평점 매트릭스 속에 숨어 있는 **잠재 요인을 추출**해 추천 예측할 수 있게 하는 기법
    - **행렬 분해**: 대규모 다차원 행렬을 SVD와 같은 **차원 감소 기법**으로 분해하는 과정에서 잠재 요인 추출

![image](https://user-images.githubusercontent.com/79856553/151136214-244a5daf-1a00-4e68-814a-5397cb77a2a7.png)

2. **잠재요인**
    - 잠재 요인이 정확히 어떤건지 특정할 수는 없음
    - ex) 영화 평점 기반 [사용자-아이템]평점행렬이라면 영화가 가지는 장르별 특성 선호도로 가정할 수 있음(사용자: 영화 장르 선호도, 아이템: 영화 장르별 특성값)

![image](https://user-images.githubusercontent.com/79856553/151137857-169b730d-ddd8-4671-b2a7-b12c2ec81ae7.png)

-> $ R(u,i) = P(u,k) \cdot Q(i,k) $

: 행렬R 행렬분해 -> 행렬P, 행렬Q, P와 Q의 내적을 통해 R값을 예측할 수 있음

- R(u,i): 사용자-아이템 평점 행렬(u: 사용자 아이디, i: 아이템 아이디)
- P(u,k): 사용자-잠재요인 행렬(u: 사용자 아이디, k: 장르별 선호도(잠재요인))
- Q(i,k): 아이템-잠재요인 행렬(i: 아이템 아이디, k: 장르별 요소(잠재요인))
    - Q.T(k,i): P와의 내적 계산 위해 전치행렬로 변환

3. **행렬 분해의 이해**
    1. 행렬분해: 다차원의 매트릭스를 저차원 매트릭스로 분해하는 기법
        - SVD, NMF 등이 있음
    2. **R = P * Q.T**
        - R행렬: MxN차원 -> M개의 사용자 행 & N개의 아이템 열
        - P행렬: MxK차원 -> M개의 사용자 & K개의 잠재요인
        - Q행렬: KxN차원 -> K개의 잠재요인 & N개의 아이템
    3. 평점데이터 $ r(u,i) = P_u * q_i^t $
        - ex) $ r(2,3) = p_2 * q_3^t $

![image](https://user-images.githubusercontent.com/79856553/151140781-9e93ee8f-f283-46ef-9566-06428df42caa.png)

4. 행렬R을 행렬P와 Q로 분해하는 방법
    1. 행렬분해는 주로 SVD: SVD는 널값이 없는 행렬에만 적용 가능
    2. **확률적 경사 하강법** or **ALS(Alternating Least Squares)** 이용해 SVD 수행

5. 확률적 경사 하강법을 이용한 행렬 분해
    1. **P와 Q행렬로 계산된 예측 R행렬 값이 실제 R행렬 값과 가장 최소의 오류를 가질 수 있도록 반복적인 비용 함수 최적화를 통해 P와 Q을 유추해내는 것**
    2. 확률적 경사 하강법 절차
        1. P와 Q를 임의의 값을 가진 행렬로 설정
        2. P와 Q.T값을 곱해 예측R행렬을 계산하고 **예측R행렬과 실제R행렬에 해당하는 오류 값 계산**
        3. 이 **오류 값을 최소화**할 수 있도록 P와 Q행렬을 적절한 값으로 각각 업데이트
        4. 만족할 만한 오류 값을 가질 때까지 2,3번 작업을 반복하면서 P와 Q값을 업데이트해 근사화
    3. **비용함수식**: 실제값과 예측값의 오류 최소화와 L2규제 고려
$$ min\sum(r(u,i)-P_u Q_i^t)^2 + \lambda(||q_i||^2 + ||p_u||^2) $$
        - $\acute{p_u} = p_u + \eta(e(u,i)*q_i-\lambda*p_u) $
        - $\acute{q_i} = p_u + \eta(e(u,i)*p_i-\lambda*q_u) $
        - $e(u,i)=r(u,i)-\hat{r}(u,i)$:u행,i열에 위치한 실제 행렬 값과 예측 행렬 값의 차이 오류
        - $\eta$: SGD 학습률
        - $\lambda$: L2 규제 계수

<font color=darkred> SGD 이용한 행렬분해 </font>

In [1]:
# R: 널 값 포함해 생성(np.NaN), P,Q는 정규분포 가진 랜던 값으로 초기화
import numpy as np

# 원본 행렬 R 생성, 분해 행렬 P와 Q 초기화, 잠재요인 차원 K는 3 설정. 
R = np.array([[4, np.NaN, np.NaN, 2, np.NaN ],
              [np.NaN, 5, np.NaN, 3, 1 ],
              [np.NaN, np.NaN, 3, 4, 4 ],
              [5, 2, 1, 2, np.NaN ]])
num_users, num_items = R.shape
K=3

# P와 Q 매트릭스의 크기를 지정하고 정규분포를 가진 random한 값으로 입력합니다. 
np.random.seed(1)
P = np.random.normal(scale=1./K, size=(num_users, K))
Q = np.random.normal(scale=1./K, size=(num_items, K))

In [2]:
# 실제 R행렬과 예측행렬의 오차 구하는 함수 - get_rmse()
from sklearn.metrics import mean_squared_error

def get_rmse(R, P, Q, non_zeros):
    error = 0
    # 두개의 분해된 행렬 P와 Q.T의 내적으로 예측 R 행렬 생성
    full_pred_matrix = np.dot(P, Q.T)
    
    # 실제 R 행렬에서 널이 아닌 값의 위치 인덱스 추출하여 실제 R 행렬과 예측 행렬의 RMSE 추출
    x_non_zero_ind = [non_zero[0] for non_zero in non_zeros]
    y_non_zero_ind = [non_zero[1] for non_zero in non_zeros]
    R_non_zeros = R[x_non_zero_ind, y_non_zero_ind]
    full_pred_matrix_non_zeros = full_pred_matrix[x_non_zero_ind, y_non_zero_ind]
      
    mse = mean_squared_error(R_non_zeros, full_pred_matrix_non_zeros)
    rmse = np.sqrt(mse)
    
    return rmse

In [3]:
# R > 0 인 행 위치, 열 위치, 값을 non_zeros 리스트에 저장. 
non_zeros = [ (i, j, R[i,j]) for i in range(num_users) for j in range(num_items) if R[i,j] > 0 ]

# steps: SGD 반복해서 업데이트할 횟수, learning_rate: SGD 학습률, r_lambda: L2 Regularization 계수
steps=1000
learning_rate=0.01
r_lambda=0.01

# SGD 기법으로 P와 Q 매트릭스를 계속 업데이트. 
for step in range(steps):
    for i, j, r in non_zeros:
        # 실제 값과 예측 값의 차이인 오류 값 구함
        eij = r - np.dot(P[i, :], Q[j, :].T)
        # Regularization을 반영한 SGD 업데이트 공식 적용
        P[i,:] = P[i,:] + learning_rate*(eij * Q[j, :] - r_lambda*P[i,:])
        Q[j,:] = Q[j,:] + learning_rate*(eij * P[i, :] - r_lambda*Q[j,:])

    rmse = get_rmse(R, P, Q, non_zeros)
    if (step % 50) == 0 :
        print("### iteration step : ", step," rmse : ", rmse)

### iteration step :  0  rmse :  3.2388050277987723
### iteration step :  50  rmse :  0.4876723101369648
### iteration step :  100  rmse :  0.1564340384819247
### iteration step :  150  rmse :  0.07455141311978046
### iteration step :  200  rmse :  0.04325226798579314
### iteration step :  250  rmse :  0.029248328780878973
### iteration step :  300  rmse :  0.022621116143829466
### iteration step :  350  rmse :  0.019493636196525135
### iteration step :  400  rmse :  0.018022719092132704
### iteration step :  450  rmse :  0.01731968595344266
### iteration step :  500  rmse :  0.016973657887570753
### iteration step :  550  rmse :  0.016796804595895633
### iteration step :  600  rmse :  0.01670132290188466
### iteration step :  650  rmse :  0.01664473691247669
### iteration step :  700  rmse :  0.016605910068210026
### iteration step :  750  rmse :  0.016574200475705
### iteration step :  800  rmse :  0.01654431582921597
### iteration step :  850  rmse :  0.01651375177473524
### iterati

In [4]:
pred_matrix = np.dot(P, Q.T)
print('예측 행렬:\n', np.round(pred_matrix, 3))

예측 행렬:
 [[3.991 0.897 1.306 2.002 1.663]
 [6.696 4.978 0.979 2.981 1.003]
 [6.677 0.391 2.987 3.977 3.986]
 [4.968 2.005 1.006 2.017 1.14 ]]


<font color=darkred> 원본 행렬과 비교해 널이 아닌 값은 큰 차이 나이 않으며, 널인 값은 새로운 예측값으로 채워짐 </font> 

## <font color = (blue)> 05. (실습1)콘텐츠 기반 필터링 실습 - TMDB 5000 영화 데이터 세트 </font>

1. **TMDB 5000 영화 데이터 세트**: 유명한 영화 데이터 정보 사이트인 IMDB의 많은 영화 중 주요 5000개 영화에 대한 메타 정보를 새롭게 가공해 캐글에서 제공하는 데이터 세트


2. **장르 속성을 이용한 영화 콘텐츠 기반 필터링**: 사용자가 특정 영화를 감상하고 그 영화를 좋아했다면 그 영화와 **비슷한 특성/속성, 구성 요소** 등을 가진 다른 영화를 추천하는 것
    - 유사성 판단 기준이 영화를 구성하는 다양한 콘텐츠: 장르, 감독, 배우, 평점, 키워드, 영화 설명

<font color=magenta> 1. 데이터 로밍 및 가공 </font>

In [7]:
# 1. 데이터 로밍 및 가공: tmdb_5000_movies.csv
import pandas as pd
import numpy as np
import warnings; warnings.filterwarnings('ignore')

movies =pd.read_csv('./tmdb-5000-movie-dataset/tmdb_5000_movies.csv')
print(movies.shape)
movies.head(1)

(4803, 20)


budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                      homepage     id  \
0  http://www.avatarmovie.com/  19995   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   

  original_title                                           overview  \
0         Avatar  In the 22nd century, a paraplegic Marine is di...   

   popularity                               production_companies  \
0  150.437577  [{"name": "Ingenious Film Partners", "id": 289...   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   

                       tagline   title  vote_average  vote_count  
0  Enter the World of Pandora.  Avatar           7.2       11800

-> 구성: 4803개 레코드 & 20개 피처

In [8]:
# 콘텐츠 기반 필터링 추천 분석에 사용할 주요 칼럼만 추출
movies_df = movies[['id','title', 'genres', 'vote_average', 'vote_count',
                 'popularity', 'keywords', 'overview']]

In [9]:
# 해당 칼럼이 어떤 형태로 되어있는지 확인
pd.set_option('max_colwidth', 100)
movies_df[['genres','keywords']][:1]

genres  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {...   

                                                                                              keywords  
0  [{"id": 1463, "name": "culture clash"}, {"id": 2964, "name": "future"}, {"id": 3386, "name": "sp...

-> 여러 개의 개별 장르 데이터 갖음. 개별 장르의 명칭은 딕셔너리의 키인 name으로 추출할 수 있음

In [10]:
# 문자열을 객체로 변환
from ast import literal_eval

movies_df['genres'] = movies_df['genres'].apply(literal_eval)
movies_df['keywords'] = movies_df['keywords'].apply(literal_eval)

In [11]:
# name 키에 해당하는 값 추출 -> 리스트 객체로 변환
movies_df['genres'] = movies_df['genres'].apply(lambda x : [ y['name'] for y in x])
movies_df['keywords'] = movies_df['keywords'].apply(lambda x : [ y['name'] for y in x])
movies_df[['genres', 'keywords']][:1]

genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   

                                                                                              keywords  
0  [culture clash, future, space war, space colony, society, space travel, futuristic, romance, spa...

<font color=magenta> 2. 장르 콘텐츠 유사도 측정 </font>
1. genres 문자열로 변경
2. 이를 CountVectorizer로 피처벡터화한 행렬 데이터 값을 코사인 유사도로 비교
3. 장르 유사도가 높은 영화 중, 평점이 높은 순으로 영화 추천

In [12]:
# CountVectorizer 이용해 피처 벡터 행렬 만들기
from sklearn.feature_extraction.text import CountVectorizer

# CountVectorizer를 적용하기 위해 공백문자로 word 단위가 구분되는 문자열로 변환. 
movies_df['genres_literal'] = movies_df['genres'].apply(lambda x : (' ').join(x))
count_vect = CountVectorizer(min_df=0, ngram_range=(1,2))
genre_mat = count_vect.fit_transform(movies_df['genres_literal'])
print(genre_mat.shape)

(4803, 276)


In [13]:
# 코사인 유사도 구하기: cosine_similairities()
from sklearn.metrics.pairwise import cosine_similarity

genre_sim = cosine_similarity(genre_mat, genre_mat)
print(genre_sim.shape)
print(genre_sim[:2])

(4803, 4803)
[[1.         0.59628479 0.4472136  ... 0.         0.         0.        ]
 [0.59628479 1.         0.4        ... 0.         0.         0.        ]]


In [14]:
# argsort()[:, ::-1]: 유사도 높은 순으로 정리된 genre_sim 객체의 비교 행 위치 인덱스 값
# 높은 순으로 정렬된 비교 행 위치 인덱스 값을 가져오고,
# 그중 0번 레코드의 비교 행 위치 인덱스 값만 샘플로 추출
genre_sim_sorted_ind = genre_sim.argsort()[:, ::-1]
print(genre_sim_sorted_ind[:1])

[[   0 3494  813 ... 3038 3037 2401]]


-> 0번 레코드의 경우 자신이 0번인 레코드를 제외하면 3494번 레코드 유사도 가장 높음. 그 다음이 813이고 가장 낲은 레코드는 2401

<font color=magenta> 3. 장르 콘텐츠 필터링을 이용한 영화 추천 </font>

In [15]:
def find_sim_movie(df, sorted_ind, title_name, top_n=10):
    
    # 인자로 입력된 movies_df DataFrame에서 'title' 컬럼이 입력된 title_name 값인 DataFrame추출
    title_movie = df[df['title'] == title_name]
    
    # title_named을 가진 DataFrame의 index 객체를 ndarray로 반환하고 
    # sorted_ind 인자로 입력된 genre_sim_sorted_ind 객체에서 유사도 순으로 top_n 개의 index 추출
    title_index = title_movie.index.values
    similar_indexes = sorted_ind[title_index, :(top_n)]
    
    # 추출된 top_n index들 출력. top_n index는 2차원 데이터 임. 
    #dataframe에서 index로 사용하기 위해서 1차원 array로 변경
    print(similar_indexes)
    similar_indexes = similar_indexes.reshape(-1)
    
    return df.iloc[similar_indexes]

In [16]:
# 영화 '대부(The Godfather)' 장르별로 유사한 영화 10개 추천
similar_movies = find_sim_movie(movies_df, genre_sim_sorted_ind, 'The Godfather',10)
similar_movies[['title', 'vote_average']]

[[2731 1243 3636 1946 2640 4065 1847 4217  883 3866]]


title  vote_average
2731                          The Godfather: Part II           8.3
1243                                    Mean Streets           7.2
3636                                   Light Sleeper           5.7
1946  The Bad Lieutenant: Port of Call - New Orleans           6.0
2640         Things to Do in Denver When You're Dead           6.7
4065                                      Mi America           0.0
1847                                      GoodFellas           8.2
4217                                            Kids           6.8
883                              Catch Me If You Can           7.7
3866                                     City of God           8.1

-> 대부 2편 가장 먼저 추천됨. 그 다음은 좋은 친구들

-> 섣불리 추천하기에 이해하기 어려운 영화들도 보임. 평점0점 영화도 존재

-> 좀 더 **많은 후보군**을 선정한 뒤 영화의 **평점에 따라 필터링**해서 최종 추천하는 방식으로 변경

In [17]:
movies_df[['title','vote_average','vote_count']].sort_values('vote_average', ascending=False)[:10]

title  vote_average  vote_count
3519          Stiff Upper Lips          10.0           1
4247     Me You and Five Bucks          10.0           2
4045     Dancer, Texas Pop. 81          10.0           1
4662            Little Big Top          10.0           1
3992                 Sardaarji           9.5           2
2386            One Man's Hero           9.3           2
2970        There Goes My Baby           8.5           2
1881  The Shawshank Redemption           8.5        8205
2796     The Prisoner of Zenda           8.4          11
3337             The Godfather           8.4        5893

-> 이름도 들어본 적 없는 영화가 유명한 영화보다 더 높은 평점. 이들은 모두 평가 횟수 매우 작음

-> **평점에 평가 횟수를 반영할 필요성**

-> 평가 횟수에 가중치가 부여된 평점
$$가중평점 = \frac{v}{v+m}*R+\frac{m}{v+m}*c$$
- v: 개별 영화에 평점 투표한 횟수
- m: 평점 부여하기 위한 최소 투표 횟수 -> 가중치 직접 조절하는 역할
- R: 개별 영화에 대한 평균 평점
- C: 전체 영화에 대한 평균 평점

In [18]:
# m 높이면 평점 투표 횟수가 많은 영화에 더 많은 가중 평점 부여 -> 60%
C = movies_df['vote_average'].mean()
m = movies_df['vote_count'].quantile(0.6)
print('C:',round(C,3), 'm:',round(m,3))

C: 6.092 m: 370.2


In [19]:
# weighted_vote_average(): 기존 평점을 새로운 가중 평점으로 변경하는 함수
percentile = 0.6
m = movies_df['vote_count'].quantile(percentile)
C = movies_df['vote_average'].mean()

def weighted_vote_average(record):
    v = record['vote_count']
    R = record['vote_average']
    
    return ( (v/(v+m)) * R ) + ( (m/(m+v)) * C )   

movies_df['weighted_vote'] = movies_df.apply(weighted_vote_average, axis=1) 

In [20]:
# 상위 10개의 영화 추출
movies_df[['title','vote_average','weighted_vote','vote_count']].sort_values('weighted_vote',
                                                                          ascending=False)[:10]

title  vote_average  weighted_vote  vote_count
1881  The Shawshank Redemption           8.5       8.396052        8205
3337             The Godfather           8.4       8.263591        5893
662                 Fight Club           8.3       8.216455        9413
3232              Pulp Fiction           8.3       8.207102        8428
65             The Dark Knight           8.2       8.136930       12002
1818          Schindler's List           8.3       8.126069        4329
3865                  Whiplash           8.3       8.123248        4254
809               Forrest Gump           8.2       8.105954        7927
2294             Spirited Away           8.3       8.105867        3840
2731    The Godfather: Part II           8.3       8.079586        3338

<font color=magenta> 최종 추천 </font>
영화'대부'

In [22]:
def find_sim_movie(df, sorted_ind, title_name, top_n=10):
    title_movie = df[df['title'] == title_name]
    title_index = title_movie.index.values
    
    # top_n의 2배에 해당하는 쟝르 유사성이 높은 index 추출 
    similar_indexes = sorted_ind[title_index, :(top_n*2)]
    similar_indexes = similar_indexes.reshape(-1)
# 기준 영화 index는 제외
    similar_indexes = similar_indexes[similar_indexes != title_index]
    
    # top_n의 2배에 해당하는 후보군에서 weighted_vote 높은 순으로 top_n 만큼 추출 
    return df.iloc[similar_indexes].sort_values('weighted_vote', ascending=False)[:top_n]

similar_movies = find_sim_movie(movies_df, genre_sim_sorted_ind, 'The Godfather',10)
similar_movies[['title', 'vote_average', 'weighted_vote']]

title  vote_average  weighted_vote
2731       The Godfather: Part II           8.3       8.079586
1847                   GoodFellas           8.2       7.976937
3866                  City of God           8.1       7.759693
1663  Once Upon a Time in America           8.2       7.657811
883           Catch Me If You Can           7.7       7.557097
281             American Gangster           7.4       7.141396
4041              This Is England           7.4       6.739664
1149              American Hustle           6.8       6.717525
1243                 Mean Streets           7.2       6.626569
2839                     Rounders           6.9       6.530427

-> 이전에 추천된 영화보다 훨씬 나은 영화 추천됨

-> 한계: 개인이 좋아하는 성향을 반영하기에는 조금 부족할 수 있음

## <font color = (blue)> 06. (실습2)아이템 기반 최근접 이웃 협업 필터링 실습 </font>

최근접 이웃 협업 필터링: 사용자 기반 & 아이템 기반
- 일반적으로 추천 정확도 뛰어난 **아이템 기반 협업 필터링**
- Grouplens사이트에서 만든 MovieLens 데이터 세트: 십만 개의 평점 정보

<font color=red> $\otimes$ 데이터세트 다운 X -> 코드&결과만 확인 </font>

<font color=magenta> 1. 데이터 가공 및 변환 </font>

In [ ]:
import pandas as pd
import numpy as np

# movies.csv: 영화에 대한 메타 정보인 title과 genres를 가지고 있는 영화 정보
movies = pd.read_csv('./ml-latest-small/movies.csv')
# ratings.csv: 사용자별로 영화에 대한 평점 매긴 데이터 세트
ratings = pd.read_csv('./ml-latest-small/ratings.csv')
print(movies.shape)
print(ratings.shape)

![image](https://user-images.githubusercontent.com/79856553/151739946-bf3c852e-6075-4d04-8564-112b9a509de4.png)

![image](https://user-images.githubusercontent.com/79856553/151740073-7453bb79-690f-4b80-bf42-e2a73b636667.png)

- 협업 필터링은 ratings.csv 데이터 세트와 같이 사용자와 아이템 간의 평점에 기반해 추천하는 시스템
- 로우(행) 레벨 형태의 원본 데이터 세트를 다음 그림과 같이 **모든 사용자를 로우로**, **모든 영화를 칼럼으로** 구성한 데이터 세트로 변경해야 함

![image](https://user-images.githubusercontent.com/79856553/151740332-fc742966-eb32-4b9c-8865-99b8dc4b2172.png)

In [ ]:
# pivot_table(): 로우 레벨의 값을 칼럼으로 변경하는데 사용
ratings = ratings[['userId', 'movieId', 'rating']]
# 로우(행): userId, 칼럼(열): movieId값, 데이터: rating칼럼 값 으로 변경됨
ratings_matrix = ratings.pivot_table('rating', index='userId', columns='movieId')
ratings_matrix.head(3)

![image](https://user-images.githubusercontent.com/79856553/151740559-659e65b3-8840-422d-a1fd-ad323297d0bd.png)

-> 칼럼명이 영화 제목(title)이 아닌 숫자 값(1,2,3..)으로 할당돼있음

-> 가독성 높이기 위해 칼럼명 movieId -> 영화명 title로 변경

-> 영화명은 movies 데이터 세트에 존재: ratings와 movies 조인해 title 칼럼 가져온 뒤, pivot_table()의 인자로 columns에 'movieId'가 아닌 'title' 입력해 title로 pivot

In [ ]:
# title 컬럼을 얻기 이해 movies 와 조인 수행
rating_movies = pd.merge(ratings, movies, on='movieId')

# columns='title' 로 title 컬럼으로 pivot 수행. 
ratings_matrix = rating_movies.pivot_table('rating', index='userId', columns='title')

# NaN 값을 모두 0 으로 변환
ratings_matrix = ratings_matrix.fillna(0)
ratings_matrix.head(3)

![image](https://user-images.githubusercontent.com/79856553/151740838-42fe734a-0483-46d5-9f7d-c72d6116b657.png)

<font color=magenta> 2. 영화 간 유사도 산출 </font>
- 영화 간 유사도: 코사인 유사도 기반. cosine_similarity()
- ratings_matrix는 userId가 기준인 행 레벨 데이터이므로 아이템(영화) 기준으로 코사인 유사도를 적용하려면 ratings_matrix가 행 기준이 영화가 되어야함
   - ratins_matrix 전치행렬: transpose()

In [ ]:
# 전치행렬 만들기
ratings_matrix_T = ratings_matrix.transpose()
ratings_matrix_T.head(3)

In [ ]:
# 코사인 유사도 구하기
from sklearn.metrics.pairwise import cosine_similarity

item_sim = cosine_similarity(ratings_matrix_T, ratings_matrix_T)

# cosine_similarity() 로 반환된 넘파이 행렬을 영화명을 매핑하여 DataFrame으로 변환
item_sim_df = pd.DataFrame(data=item_sim, index=ratings_matrix.columns,
                          columns=ratings_matrix.columns)
print(item_sim_df.shape)
item_sim_df.head(3)

-> 유사도행렬 item_sim 생성됨

In [ ]:
# 유사도 행렬을 df로 변환한 item_sim_df 이용해 영화 '대부'와 유사도 높은 상위 6개 영화 추출
item_sim_df["Godfather, The (1972)"].sort_values(ascending=False)[:6]

- 콘텐츠 기반 필터링과 차이점
: 장르가 완전히 다른 영화도 유사도가 매우 높게 나타남
- 개인적 취향 반영X, 영화간의 유사도만 가지고 추천해줌

<font color=magenta> 3. 아이템 기반 최근접 이웃 협업 필터링으로 **개인화된 영화 추천** </font>
- 아이템 기반 협업 필터링에서 개인화된 예측 평점 구하기
$$\hat{R}_{u,i} =\frac{ \sum N(S_{i,N} * R_{u,N})}{ \sum N(|S_{i,N}|)}$$
- $\hat{R}_{u,i}$: 사용자 u, 아이템 i의 개인화된 예측 평점 값
- $S_{i,N}$: 아이템 i와 가장 유사도 높은 Top_N개 아이템의 유사도 벡터
- $R_{u,N}$: 사용자 u의 아이템 i와 가장 유사도 높은 Top_N개 아이템에 대한 실제 평점 벡터
- N: 아이템의 최근접 이웃 범위 계수. 특정 아이템과 유사도가 가장 높은 Top-N개의 다른 아이템 추출하는 데 사용됨

- 영화 간 유사도 가지는 item_sim_df
- 사용자-영화 평점 df인 ratings_matrix
- 두 가지 활용해 사용자별로 최적화된 평점 스코어 예측하는 함수 만들기 -> predict_rating()
    - 인자로 넘파이 행렬 변환된 rating_matrix와 item_sim_df를 입력받아, 
    - $\hat{R}_{u,i}$ 식으로 개인화된 예측 평점 계산

In [ ]:
# 사용자별로 최적화된 평점 스코어 예측하는 함수
def predict_rating(ratings_arr, item_sim_arr ):
    ratings_pred = ratings_arr.dot(item_sim_arr)/ np.array([np.abs(item_sim_arr).sum(axis=1)])
    return ratings_pred

In [ ]:
# 개인화된 예측 평점 구하기
ratings_pred = predict_rating(ratings_matrix.values , item_sim_df.values)
ratings_pred_matrix = pd.DataFrame(data=ratings_pred, index= ratings_matrix.index,
                                   columns = ratings_matrix.columns)
ratings_pred_matrix.head(3)

In [ ]:
# 예측 결과가 실제 평점과 얼마나 차이가 있는지 확인
from sklearn.metrics import mean_squared_error

# 사용자가 평점을 부여한 영화에 대해서만 예측 성능 평가 MSE 를 구함. 
def get_mse(pred, actual):
    # Ignore nonzero terms.
    pred = pred[actual.nonzero()].flatten()
    actual = actual[actual.nonzero()].flatten()
    return mean_squared_error(pred, actual)

print('아이템 기반 모든 인접 이웃 MSE: ', get_mse(ratings_pred, ratings_matrix.values ))

- MSE 결과 약 9.89나옴
- 실제값과 예측값은 서로 스케일이 다르기 때문에 MSE 클 수도 있음. 중요한건 감소시키는 방향으로 개선하는 것.


-> 특정 영화와 **가장 비슷한 유사도를 가지는 영화에 대해서만** 유사도 벡터를 적용하는 함수로 변경
- **predict_rating_topsim()**
    - **N인자**를 가지고있어 TOP-N 유사도를 가지는 영화 유사도벡터만 예측값 계산하는데 적용
    - 개별 예측값을 구하기위해 행,열별로 for루프를 반복 수행하기 때문에 **수행시간 오래걸림**

In [ ]:
def predict_rating_topsim(ratings_arr, item_sim_arr, n=20):
    # 사용자-아이템 평점 행렬 크기만큼 0으로 채운 예측 행렬 초기화
    pred = np.zeros(ratings_arr.shape)

    # 사용자-아이템 평점 행렬의 열 크기만큼 Loop 수행. 
    for col in range(ratings_arr.shape[1]):
        # 유사도 행렬에서 유사도가 큰 순으로 n개 데이터 행렬의 index 반환
        top_n_items = [np.argsort(item_sim_arr[:, col])[:-n-1:-1]]
        # 개인화된 예측 평점을 계산
        for row in range(ratings_arr.shape[0]):
            pred[row, col] = item_sim_arr[col, :][top_n_items].dot(ratings_arr[row, :][top_n_items].T) 
            pred[row, col] /= np.sum(np.abs(item_sim_arr[col, :][top_n_items]))        
    return pred

In [ ]:
# 1. predict_rating_topsim() 함수 이용해 예측 평점 계산
# 2. 실제 평점과의 MSE 구하기
ratings_pred = predict_rating_topsim(ratings_matrix.values , item_sim_df.values, n=20)
print('아이템 기반 인접 TOP-20 이웃 MSE: ', get_mse(ratings_pred, ratings_matrix.values ))


# 계산된 예측 평점 데이터는 DataFrame으로 재생성
ratings_pred_matrix = pd.DataFrame(data=ratings_pred, index= ratings_matrix.index,
                                   columns = ratings_matrix.columns)

- MSE 약3.69
- 이전보다 향상됨

In [ ]:
# 9번 사용자가 평점 준 영화 높은 순으로 나열
user_rating_id = ratings_matrix.loc[9, :]
user_rating_id[ user_rating_id > 0].sort_values(ascending=False)[:10]

![image](https://user-images.githubusercontent.com/79856553/151984419-19c5a731-e4d6-456a-b268-d6bb85947d97.png)

-> 전반적으로 흥행성 좋은 영화에 높은 평점 줌

In [ ]:
# 사용자가 이미 평점을 준 영화를 제외하고 추천할 수 있도록 
# 평점 주지 않은 영화를 리스트 객체로 반환하는 함수 생성
def get_unseen_movies(ratings_matrix, userId):
    # userId로 입력받은 사용자의 모든 영화정보 추출하여 Series로 반환함. 
    # 반환된 user_rating 은 영화명(title)을 index로 가지는 Series 객체임. 
    user_rating = ratings_matrix.loc[userId,:]
    
    # user_rating이 0보다 크면 기존에 관람한 영화임. 대상 index를 추출하여 list 객체로 만듬
    already_seen = user_rating[ user_rating > 0].index.tolist()
    
    # 모든 영화명을 list 객체로 만듬. 
    movies_list = ratings_matrix.columns.tolist()
    
    # list comprehension으로 already_seen에 해당하는 movie는 movies_list에서 제외함. 
    unseen_list = [ movie for movie in movies_list if movie not in already_seen]
    
    return unseen_list

In [ ]:
# 사용자가 영화의 평점을 주지 않은 추천 대상 영화 정보와
# predict_rating_topsim()에서 추출한 사용자별 아이템 유사도에 기반한 예측 평점 데이터 세트 이용해
# 최종적으로 사용자에게 영화 추천하는 함수
def recomm_movie_by_userid(pred_df, userId, unseen_list, top_n=10):
    # 예측 평점 DataFrame에서 사용자id index와 unseen_list로 들어온 영화명 컬럼을 추출하여
    # 가장 예측 평점이 높은 순으로 정렬함. 
    recomm_movies = pred_df.loc[userId, unseen_list].sort_values(ascending=False)[:top_n]
    return recomm_movies
    
# 사용자가 관람하지 않는 영화명 추출   
unseen_list = get_unseen_movies(ratings_matrix, 9)

# 아이템 기반의 인접 이웃 협업 필터링으로 영화 추천 
recomm_movies = recomm_movie_by_userid(ratings_pred_matrix, 9, unseen_list, top_n=10)

# 평점 데이타를 DataFrame으로 생성. 
recomm_movies = pd.DataFrame(data=recomm_movies.values,index=recomm_movies.index,columns=['pred_score'])
recomm_movies

![image](https://user-images.githubusercontent.com/79856553/151985044-b2c5702d-edb7-4a75-b46e-57a5602edfd5.png)

-> 다양하지만 높은 흥행성 가진 작품 추천됨

## <font color = (blue)> 07. (실습3)행렬 분해를 이용한 잠재 요인 협업 필터링 실습 </font>

1. SVD: 일반적으로 행렬 분해해 자주 사용됨
    - 사용자-사이템 평점 행렬에는 사용자가 평점 매기지 않은 널 데이터가 많기 때문에 **주로 SGD나 ALS기반의 행렬 분해 이용**
    - 여기서는 SGD기반 이용
    
    
2. 4절의 "잠재 요인 협업 필터링-확률적 경사 하강법을 이용한 행렬 분해" 예제 코드 사용
    - 단, 이번에는 행렬 분해 로직 부분을 함수로 만듦 -> matrix_factorization()
    - **matrix_factorization**(R, K, steps=200, learning_rate=0.01, r_lambda=0.01)
        - R: 원본 사용자-아이템 평점 행렬
        - K: 잠재 요인의 차원의 수
        - steps: SGD의 반복 횟수
        - learning rate: 학습률
        - r_lambda: L2 규제 계수

3. Suprise 이용한 추천 시스템 구축

In [25]:
# 경사 하강법을 이횽한 행렬 분해
import numpy as np
from sklearn.metrics import mean_squared_error

def get_rmse(R, P, Q, non_zeros):
    error = 0
    # 두개의 분해된 행렬 P와 Q.T의 내적 곱으로 예측 R 행렬 생성
    full_pred_matrix = np.dot(P, Q.T)
    
    # 실제 R 행렬에서 널이 아닌 값의 위치 인덱스 추출하여 실제 R 행렬과 예측 행렬의 RMSE 추출
    x_non_zero_ind = [non_zero[0] for non_zero in non_zeros]
    y_non_zero_ind = [non_zero[1] for non_zero in non_zeros]
    R_non_zeros = R[x_non_zero_ind, y_non_zero_ind]
    
    full_pred_matrix_non_zeros = full_pred_matrix[x_non_zero_ind, y_non_zero_ind]
      
    mse = mean_squared_error(R_non_zeros, full_pred_matrix_non_zeros)
    rmse = np.sqrt(mse)
    
    return rmse

In [24]:
def matrix_factorization(R, K, steps=200, learning_rate=0.01, r_lambda = 0.01):
    num_users, num_items = R.shape
    # P와 Q 매트릭스의 크기를 지정하고 정규분포를 가진 랜덤한 값으로 입력합니다. 
    np.random.seed(1)
    P = np.random.normal(scale=1./K, size=(num_users, K))
    Q = np.random.normal(scale=1./K, size=(num_items, K))

    break_count = 0
       
    # R > 0 인 행 위치, 열 위치, 값을 non_zeros 리스트 객체에 저장. 
    non_zeros = [ (i, j, R[i,j]) for i in range(num_users) for j in range(num_items) if R[i,j] > 0 ]
   
    # SGD기법으로 P와 Q 매트릭스를 계속 업데이트. 
    for step in range(steps):
        for i, j, r in non_zeros:
            # 실제 값과 예측 값의 차이인 오류 값 구함
            eij = r - np.dot(P[i, :], Q[j, :].T)
            # Regularization을 반영한 SGD 업데이트 공식 적용
            P[i,:] = P[i,:] + learning_rate*(eij * Q[j, :] - r_lambda*P[i,:])
            Q[j,:] = Q[j,:] + learning_rate*(eij * P[i, :] - r_lambda*Q[j,:])
       
        rmse = get_rmse(R, P, Q, non_zeros)
        if (step % 10) == 0 :
            print("### iteration step : ", step," rmse : ", rmse)
            
    return P, Q

In [ ]:
# 영화 평점 데이터 새롭게 df 로딩 & 사용자-아이템 평점 행렬 만들기
import pandas as pd
import numpy as np

movies = pd.read_csv('./ml-latest-small/movies.csv')
ratings = pd.read_csv('./ml-latest-small/ratings.csv')
ratings = ratings[['userId', 'movieId', 'rating']]
ratings_matrix = ratings.pivot_table('rating', index='userId', columns='movieId')

# title 컬럼을 얻기 이해 movies 와 조인 수행
rating_movies = pd.merge(ratings, movies, on='movieId')

# columns='title' 로 title 컬럼으로 pivot 수행. 
ratings_matrix = rating_movies.pivot_table('rating', index='userId', columns='title')

In [ ]:
# matrix_factorization() 함수 이용해 행렬 분해
# 시간 오래걸리므로 steop(SGD 반복횟수) 200으로 지정
P, Q = matrix_factorization(ratings_matrix.values, K=50, steps=200, learning_rate=0.01, r_lambda = 0.01)
pred_matrix = np.dot(P, Q.T)

![image](https://user-images.githubusercontent.com/79856553/151985892-4a22f0dc-eb76-47ef-8298-c09548aead63.png)

In [ ]:
# 사용자-아이템 평점 행렬을 영화 타이틀을 칼럼명으로 가지는 df로 변경
ratings_pred_matrix = pd.DataFrame(data=pred_matrix, index= ratings_matrix.index,
                                   columns = ratings_matrix.columns)

ratings_pred_matrix.head(3)

![image](https://user-images.githubusercontent.com/79856553/151986065-5e57b972-e43e-4241-b04e-3b8109f16e83.png)

In [ ]:
# 사용자가 관람하지 않는 영화명 추출   
unseen_list = get_unseen_movies(ratings_matrix, 9)

# 아이템 기반의 인접 이웃 협업 필터링으로 영화 추천 
recomm_movies = recomm_movie_by_userid(ratings_pred_matrix, 9, unseen_list, top_n=10)

# 평점 데이타를 DataFrame으로 생성. 
recomm_movies = pd.DataFrame(data=recomm_movies.values,index=recomm_movies.index,columns=['pred_score'])
recomm_movies

![image](https://user-images.githubusercontent.com/79856553/151986261-32912bab-5ddf-48f6-9c1c-8df9da2aa345.png)

## 08. 파이썬 추천 시스템 패키지 - Surprise

1. Surprise 패키지
    - 파이썬 기반의 **추천 시스템 구축**을 위한 전용 패키지 중 하나
    - 파이썬 기반에서 사이킷런과 유사한 API와 프레임워크 제공
    - conda나 pip통해 설치
    

2. Suprise 패키지 **장점**
    - **다양한 추천 알고리즘**: 사용자 또는 아이템 기반 최근접 이웃 필터링, SVD, SVD++, NMF 기반의 잠재 요인 협업 필터링을 쉽게 적용해 추천 시스템 구축할 수 있음
    - Surprise의 핵심 API는 사이킷런의 핵심 API와 유사한 API명 가짐
        - fit(), predict(), train_test_split(), cross_validate(), GridSearchCV클래스

![image](https://user-images.githubusercontent.com/79856553/151738466-7ecd0e79-02c4-45e0-9b8e-63dcdbb4a1fe.png)

3. Surprise를 이용한 추천 시스템 구축

In [27]:
# Surprise 관련 모듈 임포트
from surprise import SVD
from surprise import Dataset 
from surprise import accuracy 
from surprise.model_selection import train_test_split

In [28]:
# 추천을 위한 데이터 세트 로딩
# -> Surprise에서 데이터 로딩은 Dataset클래스를 이용해서만 가능
data = Dataset.load_builtin('ml-100k') 
trainset, testset = train_test_split(data, test_size=.25, random_state=0) 

Dataset ml-100k could not be found. Do you want to download it? [Y/n] y
Trying to download dataset from http://files.grouplens.org/datasets/movielens/ml-100k.zip...
Done! Dataset ml-100k has been saved to C:\Users\USER/.surprise_data/ml-100k


- ml-100k: 무비렌즈 사이트에서 제공하는 과거 버전 데이터 세트. 10만개 평점 데이터
- 로컬 디렉터리에 데이터 저장된 후에는 사이트에 접속하지 않고 저장된 데이터 로딩
    - Dataset.load_builtin('ml-100k')
- 주의)과거 버전 데이터 세트: 칼럼 분리문자가 콤마가 아닌 탭(\t)

In [30]:
# 알고리즘 객체 생성 & 학습
algo = SVD()
algo.fit(trainset) 

In [31]:
# test(): 입력된 데이터 세트에 대해 추천 데이터 세트 만들어줌
# predict(): 개별 사용자와 영화에 대한 추천 평점 반환
predictions = algo.test( testset )
print('prediction type :',type(predictions), ' size:',len(predictions))
print('prediction 결과의 최초 5개 추출')
predictions[:5]

prediction type : <class 'list'>  size: 25000
prediction 결과의 최초 5개 추출


[Prediction(uid='120', iid='282', r_ui=4.0, est=3.386932093447826, details={'was_impossible': False}),
 Prediction(uid='882', iid='291', r_ui=4.0, est=3.8147246721394614, details={'was_impossible': False}),
 Prediction(uid='535', iid='507', r_ui=5.0, est=3.9161230361701462, details={'was_impossible': False}),
 Prediction(uid='697', iid='244', r_ui=5.0, est=3.7186230640100977, details={'was_impossible': False}),
 Prediction(uid='751', iid='385', r_ui=4.0, est=3.351433998597597, details={'was_impossible': False})]

##### * test() 호출 결과
- **파이썬 리스트**
- 크기: 입력 인자 데이터 세트 크기와 같은 25,000개
- 호출 결과로 반환된 리스트 객체는 25,000개의 Prediction 객체를 내부에 가지고 있음
    - **Prediction객체**: surprise 패키지에서 제공하는 데이터 타입
        - 개별 사용자 아이디, 영화 아이디, 실제 평점 정보에 기반해 Surprise의 추천 예측 평점 데이터를 튜플 형태로 가지고 있음
    - **Prediction 객체의 details 속성**: 내부 처리 시 추천 예측할 수 없는 경우 로그 남기는데 사용
         - details={'was_impossible': True}: 예측값 생성할 수 없는 데이터

In [32]:
# Prediction 객체에서 uid, iid, est 속성 추출
[ (pred.uid, pred.iid, pred.est) for pred in predictions[:3] ]

[('120', '282', 3.386932093447826),
 ('882', '291', 3.8147246721394614),
 ('535', '507', 3.9161230361701462)]

In [33]:
# 사용자 아이디, 아이템 아이디는 문자열로 입력해야 함. 
uid = str(196)
iid = str(302)
pred = algo.predict(uid, iid)
print(pred)

user: 196        item: 302        r_ui = None   est = 4.17   {'was_impossible': False}


##### * predict()
- 개별 사용자의 아이템에 대한 추천 평점 예측 -> est로 반환
- **test()메서드는** 입력 데이터 세트의 모든 사용자와 아이템 아이디에 대해서 **predict()를 반복적으로 수행한 결과**라고 생각하면 이해하기 쉬움


In [34]:
# 추천 예측 평점과 실제 평점과의 차이 평가
# Surpreise accuracy 모듈은 RMSE, MSE 등 제공
accuracy.rmse(predictions)

RMSE: 0.9482


0.9481869292906402

4. Surprise 주요 모듈 소개
    - **Dataset**: Suprise는 user_id, item_id, rating 데이터가 로우 레벨로 된 데이터 세트만 적용할 수 있음
    - 데이터 세트의 칼럼 순서가 사용자 아이디, 아이템 아이디, 평점 순으로 돼 있어야 함
    1. Dataset.load_builtin: 무비렌즈 아카이브 FTP 서버에서 무비렌즈 데이터 다운
    2. Dataset.load_from_file: OS 파일에서 데이터 로딩할 때 사용
        - 입력 파라미터로 os파일명, reader로 파일의 포맷 지정
    3. Dataset.load_from_df: 판다스의 DataFrame에서 데이터 로딩
        - df는 반드시 3개의 칼럼 user_id, item_df, rating 순
        - 입력 파라미터로 df객체, reader로 파일 포맷 지정

##### <font color=magenta> OS 파일 데이터를 Surprise 데이터 세트로 로딩 </font>
<font color=red>-> Grouplens 사이트 접속 불가(ratings.csv, movies.csv)</font>

Surprise에 OS 파일 로딩 시 주의점
- 데이터 파일에 칼럼명 가지는 헤더 문자열 있으면 안됨
- ratings.csv 파일 칼럼명 헤더로 가짐 -> to_csv() 이용해 헤더 삭제

In [ ]:
import pandas as pd

ratings = pd.read_csv('./ml-latest-small/ratings.csv')
# ratings_noh.csv 파일로 unload 시 index 와 header를 모두 제거한 새로운 파일 생성.  
ratings.to_csv('./ml-latest-small/ratings_noh.csv', index=False, header=False)

1) Reader 클래스 이용해 데이터 파일의 파싱 포맷 정의


2) Reader 클래스 생성자에 각 필드의 칼럼명, 칼럼 분리문자, 최소~최대 평점 입력해 객체 생성


3) load_from_file()로 생성된 Reader객체 참조에 데이터 파일 파싱하면서 로딩

In [ ]:
from surprise import Reader

reader = Reader(line_format='user item rating timestamp', sep=',', rating_scale=(0.5, 5))
data=Dataset.load_from_file('./ml-latest-small/ratings_noh.csv',reader=reader)

** Reader 클래스 주요 생성 파라미터
1) **line_format(string)**: 칼럼 순서대로 나열. 입력된 문자열 공백으로 분리해 칼럼 인식

2) **sep(char)**: 칼럼 분리하는 분리자. 디폴트는 \t. 판다스 df에서 입력받을 경우 기재할 필요 없음

3) **rating_scale(tuple, optional)**: 평점 값의 최소~최대 평점 설정. 디폴트 (1,5)

In [ ]:
# SVD 행렬 분해 기법 이용해 추천 예측
trainset, testset = train_test_split(data, test_size=.25, random_state=0)

# 수행시마다 동일한 결과 도출을 위해 random_state 설정 
algo = SVD(n_factors=50, random_state=0)

# 학습 데이터 세트로 학습 후 테스트 데이터 세트로 평점 예측 후 RMSE 평가
algo.fit(trainset) 
predictions = algo.test( testset )
accuracy.rmse(predictions)

##### <font color=magenta> 판다스 DataFrame에서 Surprise 데이터 세트로 로딩 </font>
- 사용자아이디, 아이템아이디, 평점 칼럼 순서 주의

In [ ]:
import pandas as pd
from surprise import Reader, Dataset

ratings = pd.read_csv('./ml-latest-small/ratings.csv') 
reader = Reader(rating_scale=(0.5, 5.0))

# ratings DataFrame 에서 컬럼은 사용자 아이디, 아이템 아이디, 평점 순서를 지켜야 합니다. 
data = Dataset.load_from_df(ratings[['userId', 'movieId', 'rating']], reader)
trainset, testset = train_test_split(data, test_size=.25, random_state=0)

algo = SVD(n_factors=50, random_state=0)
algo.fit(trainset) 
predictions = algo.test( testset )
accuracy.rmse(predictions)

5. Surprise 추천 알고리즘 클래스
    1. **SVD**: 행렬 분해를 통한 잠재 요인 협업 필터링을 위한 SVD 알고리즘
        1. SVD 클래스의 입력 파라미터
            - **n_factors**: 잠재요인 K의 개수. **디폴트 100**. 커질수록 정확도 높음/과적합 주의
            - **n_epochs**: SGD 수행 시 반복 횟수. **디폴트 20**
            - **biased(bool)**: 베이스라인 사용자 편향 적용 여부. **디폴트 True**
        2. Surprise SVD의 비용 함수: 사용자 베이스라인 편향성을 감안한 평점 예측에 Regularization 적용한 것         
    2. **KNNBasic**: 최근접 이웃 협업 필터링을 위한 KNN 알고리즘
    3. **BaselineOnly**: 사용자 Bias와 아이템 Bias를 감안한 SGD 베이스라인 알고리즘

![image](https://user-images.githubusercontent.com/79856553/151997969-625c0cfd-e0b9-4cbd-a63d-e3707b6b6e04.png)

-> 추천 알고리즘 예측 성능 벤치마크 결과
- SVD++ 알고리즘이 RMSE, MAE 성적 가장 좋지만, 상대적으로 시간 오래걸림
- SVD, K-NN Baseline 다음으로 성능 평가 수치 좋음
    - Baseline: 각 개인이 평점을 부여하는 성향을 반영해 평점을 계산하는 방식
    - K-NN 자체는 성능 상대적으로 떨어지지만 Baseline 결합해 대폭 향상

6. 베이스라인 평점
: 개인의 성향을 반영해 아이템 평가에 편향성 요소를 반영하여 평점을 부과하는 것
    - 어떤 사람은 엄격한 잣대로 평가하고, 어떤 사람은 수고한 강사를 생각해 별로였어도 좋은 점수를 주기도 함
- 베이스 라인 평점 공식: 전체 평균 평점 + 사용자 편향 점수 + 아이템 평향 점수
    - 전체 평균 평점: 모든 사용자 아이템에 대한 평점 평균한 값
    - 사용자 편향 점수: 사용자별 아이템 평점 평균 값 - 전체 평균 평점
    - 아이템 평향 점수: 아이템별 평점 평균 값 - 전체 평균 평점

![image](https://user-images.githubusercontent.com/79856553/151999327-703b3062-b156-494b-97d8-4b2c23aa0ffe.png)

-> 영화 평점 베이스라인 평점 고려해 적용
- 전체 평균 평점: 3.5, 어벤저스3편의 평균 평점: 4.2 일 때
- 전체 평균 평점 = 3.5
- 사용자 편향 점수 = 3.0-3.5=-0.5
- 아이템 편향 점수 = 4.2-3.5=0.7
- **베이스라인 평점 = 3.5 - 0.5 + 0.7 = 3.7**

7. 교차 검증과 하이퍼 파라미터 튜닝
: Surprise는 교차 검증과 하이퍼 파라미터 튜닝 위해 사이킷런과 유사한 cross_validate()와 gridSearchCV 클래스 제공
   1. **교차 검증 cross_validate()**
       - surprise.model_selection 모듈 내에 존재
       - 폴드된 데이터 세트의 개수와 성능 측정 방법 명시해 교차 검증 수행

In [ ]:
# cross_validate() 이용해 ratings.csv를 df로 로딩한 데이터 교차검증 수행, 성능평가
from surprise.model_selection import cross_validate 

# Pandas DataFrame에서 Surprise Dataset으로 데이터 로딩 
ratings = pd.read_csv('./ml-latest-small/ratings.csv') # reading data in pandas df
reader = Reader(rating_scale=(0.5, 5.0))
data = Dataset.load_from_df(ratings[['userId', 'movieId', 'rating']], reader)

algo = SVD(random_state=0) 
cross_validate(algo, data, measures=['RMSE', 'MAE'], cv=5, verbose=True) 

![image](https://user-images.githubusercontent.com/79856553/152001135-0f857136-25b4-4429-abeb-98699a9d0b8a.png)

B. **하이퍼 파라미터 튜닝 GridSearchCV**
- 사이킷런에서와 유사하게 교차 검증 통한 하이퍼 파라미터 최적화 수행
- SVD의 경우 주로 점진적 하강 방식의 반복횟수 지정하는 n_epochs와 SVD의 잠재 요인 K의 크기를 지정하는 n_factors 튜닝

In [ ]:
from surprise.model_selection import GridSearchCV

# 최적화할 파라미터들을 딕셔너리 형태로 지정. 
param_grid = {'n_epochs': [20, 40, 60], 'n_factors': [50, 100, 200] }

# CV를 3개 폴드 세트로 지정, 성능 평가는 rmse, mse 로 수행 하도록 GridSearchCV 구성
gs = GridSearchCV(SVD, param_grid, measures=['rmse', 'mae'], cv=3)
gs.fit(data)

# 최고 RMSE Evaluation 점수와 그때의 하이퍼 파라미터
print(gs.best_score['rmse'])
print(gs.best_params['rmse'])

![image](https://user-images.githubusercontent.com/79856553/152001580-267fa6eb-4b53-4d9e-9c06-69203f4d8b96.png)

8. Surprise를 이용한 개인화 영화 추천 시스템 구축
- ratings.csv 데이터를 학습 데이터와 테스트 데이터로 분리하지 않고 전체 학습 데이터로 사용
- 데이터 세트 전체를 학습 데이터로 사용하려면 **DatasetAutoFolds 클래스** 이용

In [ ]:
from surprise.dataset import DatasetAutoFolds

reader = Reader(line_format='user item rating timestamp', sep=',', rating_scale=(0.5, 5))
# DatasetAutoFolds 클래스를 ratings_noh.csv 파일 기반으로 생성. 
data_folds = DatasetAutoFolds(ratings_file='./ml-latest-small/ratings_noh.csv', reader=reader)

#전체 데이터를 학습데이터로 생성함. 
trainset = data_folds.build_full_trainset()

In [ ]:
# SVD 학습 수행
algo = SVD(n_epochs=20, n_factors=50, random_state=0)
algo.fit(trainset)

In [ ]:
# 영화에 대한 상세 속성 정보 DataFrame로딩
movies = pd.read_csv('./ml-latest-small/movies.csv')

# userId=9 의 movieId 데이터 추출하여 movieId=42 데이터가 있는지 확인. 
movieIds = ratings[ratings['userId']==9]['movieId']
if movieIds[movieIds==42].count() == 0:
    print('사용자 아이디 9는 영화 아이디 42의 평점 없음')

print(movies[movies['movieId']==42])

In [ ]:
# 학습된 SVD 객체에서 predict() 메서드 내에 userId와 movieId값 입력 -> 문자열
uid = str(9)
iid = str(42)

pred = algo.predict(uid, iid, verbose=True)

In [ ]:
# 아이디 9인 사용자가 아직 평점 매기지 않은 영화 정보 반환
def get_unseen_surprise(ratings, movies, userId):
    #입력값으로 들어온 userId에 해당하는 사용자가 평점을 매긴 모든 영화를 리스트로 생성
    seen_movies = ratings[ratings['userId']== userId]['movieId'].tolist()
    
    # 모든 영화들의 movieId를 리스트로 생성. 
    total_movies = movies['movieId'].tolist()
    
    # 모든 영화들의 movieId중 이미 평점을 매긴 영화의 movieId를 제외하여 리스트로 생성
    unseen_movies= [movie for movie in total_movies if movie not in seen_movies]
    print('평점 매긴 영화수:',len(seen_movies), '추천대상 영화수:',len(unseen_movies), \
          '전체 영화수:',len(total_movies))
    
    return unseen_movies

unseen_movies = get_unseen_surprise(ratings, movies, 9)

-> 결과) 전체 9742개 영화 중 46개만 평점 매김
- 추천 대상 영화 9696개
- SVD 이용해 높은 예측 평점 가진 순으로 영화 추천

In [ ]:
def recomm_movie_by_surprise(algo, userId, unseen_movies, top_n=10):
    # 알고리즘 객체의 predict() 메서드를 평점이 없는 영화에 반복 수행한 후 결과를 list 객체로 저장
    predictions = [algo.predict(str(userId), str(movieId)) for movieId in unseen_movies]
    
    # predictions list 객체는 surprise의 Predictions 객체를 원소로 가지고 있음.
    # [Prediction(uid='9', iid='1', est=3.69), Prediction(uid='9', iid='2', est=2.98),,,,]
    # 이를 est 값으로 정렬하기 위해서 아래의 sortkey_est 함수를 정의함.
    # sortkey_est 함수는 list 객체의 sort() 함수의 키 값으로 사용되어 정렬 수행.
    def sortkey_est(pred):
        return pred.est
    
    # sortkey_est( ) 반환값의 내림 차순으로 정렬 수행하고 top_n개의 최상위 값 추출.
    predictions.sort(key=sortkey_est, reverse=True)
    top_predictions= predictions[:top_n]
    
    # top_n으로 추출된 영화의 정보 추출. 영화 아이디, 추천 예상 평점, 제목 추출
    top_movie_ids = [ int(pred.iid) for pred in top_predictions]
    top_movie_rating = [ pred.est for pred in top_predictions]
    top_movie_titles = movies[movies.movieId.isin(top_movie_ids)]['title']
    top_movie_preds = [ (id, title, rating) for id, title, rating in zip(top_movie_ids, top_movie_titles, top_movie_rating)]
    
    return top_movie_preds

unseen_movies = get_unseen_surprise(ratings, movies, 9)
top_movie_preds = recomm_movie_by_surprise(algo, 9, unseen_movies, top_n=10)
print('##### Top-10 추천 영화 리스트 #####')

for top_movie in top_movie_preds:
    print(top_movie[1], ":", top_movie[2])

![image](https://user-images.githubusercontent.com/79856553/152002775-0bef22ab-3f88-48ce-8f6f-ab353c8de27e.png)

## 09. 정리

1. 추천 시스템의 대표적인 방식: 콘텐츠 기반 필터링 & 협업 필터링


2. **콘텐츠 기반 필터링**: 아이템을 구성하는 여러 가지 콘텐츠 중 사용자가 좋아하는 콘텐츠를 필터링하여 이에 맞는 아이템 추천하는 방식
    1. 사용자가 좋아하는 영화 분석한 뒤, 그 **영화 콘텐츠와 유사한 다른 영화 추천**
    
    
3. **협업 필터링** - 최근접 이웃 협업 필터링 & 잠재 요인 협업 필터링
    1. **최근접 이웃 협업 필터링** - 사용자 기반 & 아이템 기반
        1. 일반적으로 아이템 기반이 정확도 더 높음
        2. **특정 아이템과 가장 근접하게 유사한 다른 아이템 추천**(아이템 기반)
        3. 유사도 기준: 사용자들의 아이템에 대한 평가를 벡터화한 값
        4. 아이템-사용자 행렬 데이터 세트 만들고, 아이템 별로 코사인 유사도 이용해 사용자 평점 피처 벡처에 따른 TOP-N 유사 아이템 추천
    2. **잠재 요인 협업 필터링**
        1. 사용자-아이템 평점 행렬 데이터에 **숨어 있는 잠재 요인을 추출**하여 사용자가 **아직 평점을 매기지 않은 아이템에 대한 평점 예측**하여 추천에 반영
        2. 다차원 사용자-아이템 평점 행렬 -(행렬분해)-> 저차원 사용자-잠재요인, 아이템-잠재요인 행렬
        3. 행렬 분해 기법은 경사 하강법으로 구현
        
        
4. **Surprise**
    - 파이썬의 추천 시스템 패키지 중 하나
    - 사이킷런과 유사한 API 지향
    - 간단한 API만 이용해 파이썬 기반에서 추천 시스템 구현해줌